src/app/libp2p_helper/src/libp2p_helper/bitswap.go

```diff
@@ -0,0 +1,268 @@
+}
+
+type bitswapAddCmd struct {
+	tag  BitswapDataTag
+	data []byte
+}
+
+type bitswapDownloadCmd struct {
+	tag     BitswapDataTag
+	rootIds []root
+}
+
+type BitswapCtx struct {
+	downloadCmds       chan bitswapDownloadCmd
+	addCmds            chan bitswapAddCmd
+	deleteCmds         chan bitswapDeleteCmd
+	engine             *bitswap.Bitswap
+	storage            codanet.BitswapStorage
+	ctx                context.Context
+	blockSink          chan blocks.Block
+	nodeDownloadParams map[cid.Cid]map[root][]NodeIndex
+	rootDownloadStates map[root]*RootDownloadState
+	deadlineChan       chan root
+	outMsgChan         chan<- *capnp.Message
+	maxBlockSize       int
+	dataConfig         map[BitswapDataTag]BitswapDataConfig
+	depthIndices       DepthIndices
+}
+
+func NewBitswapCtx(ctx context.Context, outMsgChan chan<- *capnp.Message) *BitswapCtx {
+	maxBlockSize := 1 << 18 // 256 KiB
+	return &BitswapCtx{
+		downloadCmds:       make(chan bitswapDownloadCmd, 100),
+		addCmds:            make(chan bitswapAddCmd, 100),
+		deleteCmds:         make(chan bitswapDeleteCmd, 100),
+		ctx:                ctx,
+		rootDownloadStates: make(map[root]*RootDownloadState),
+		nodeDownloadParams: make(map[cid.Cid]map[root][]NodeIndex),
+		blockSink:          make(chan blocks.Block, 100),
+		deadlineChan:       make(chan root, 100),
+		outMsgChan:         outMsgChan,
+		maxBlockSize:       maxBlockSize,
+		dataConfig: map[BitswapDataTag]BitswapDataConfig{
+			BlockBodyTag: {
+				maxSize:         1 << 26, // 64 MiB
+				downloadTimeout: time.Minute * 10,
+			},
+		},
+		depthIndices: MkDepthIndices(LinksPerBlock(maxBlockSize), math.MaxInt32),
+	}
+}
+
+func announceNewRootBlock(engine *bitswap.Bitswap, statusStorage codanet.BitswapStorage, bs map[BitswapBlockLink][]byte, root BitswapBlockLink) error {
+	err := statusStorage.SetStatus(root, codanet.Partial)
+	if err != nil {
+		return err
+	}
+
+	for h, b := range bs {
+		bitswapLogger.Debugf("Publishing block %s (%d bytes)", codanet.BlockHashToCidSuffix(h), len(b))
+		block, _ := blocks.NewBlockWithCid(b, codanet.BlockHashToCid(h))
+		err = engine.HasBlock(block)
+		if err != nil {
+			return err
+		}
+	}
+	return statusStorage.SetStatus(root, codanet.Full)
+}
+
+func (bs *BitswapCtx) deleteRoot(root BitswapBlockLink) error {
+	err := bs.storage.SetStatus(root, codanet.Deleting)
+	if err != nil {
+		return err
+	}
+	FreeRoot(bs, root)
+	allDescedants := []BitswapBlockLink{root}
+	for i := 0; i < len(allDescedants); i++ {
+		block := allDescedants[i]
+		err := bs.storage.ViewBlock(block, func(b []byte) error {
+			links, _, err := ReadBitswapBlock(b)
+			if err == nil {
+				for _, l := range links {
+					var l2 BitswapBlockLink
+					copy(l2[:], l[:])
+					allDescedants = append(allDescedants, l2)
+				}
+			}
+			return err
+		})
+		if err != nil && err != blockstore.ErrNotFound {
+			return err
+		}
+	}
+	return bs.storage.DeleteBlocks(allDescedants)
+}
+
+func FreeRoot(bs BitswapState, root root) {
```

Comment:
This name confused me for a bit when reading the code (I was thinking this function would free the root its children from block storage). Maybe a better name could be `FreeRootDownloadState` or `ClearDownloadState`?

Reply:
Renamed to `ClearRootDownloadState`